File: discord-bot-fleet/src/tools/web_search.ts
// web_search — minimal web search via DuckDuckGo HTML endpoint.
// No API key required. Returns top 5 results with title/url/snippet.

import type { LLMTool } from '../llm.js';

export const webSearchTool: LLMTool = {
  type: 'function',
  function: {
    name: 'web_search',
    description: 'Search the web for current information. Returns up to 5 results with title, URL, and a short snippet. Use this when you need up-to-date info beyond your training data.',
    parameters: {
      type: 'object',
      properties: {
        query: { type: 'string', description: 'Search query' },
      },
      required: ['query'],
    },
  },
};

export async function webSearchHandler(args: { query: string }): Promise<string> {
  const url = `https://html.duckduckgo.com/html/?q=${encodeURIComponent(args.query)}`;
  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), 10000);
  try {
    const res = await fetch(url, {
      signal: controller.signal,
      headers: { 'User-Agent': 'Mozilla/5.0 (compatible; DiscordBot/1.0)' },
    });
    if (!res.ok) return JSON.stringify({ error: `HTTP ${res.status}` });
    const html = await res.text();
    // Parse result blocks: <a class="result__a" href="...">title</a> + <a class="result__snippet">...</a>
    const results: Array<{ title: string; url: string; snippet: string }> = [];
    const linkRe = /<a class="result__a"[^>]*href="([^"]+)"[^>]*>([\s\S]*?)<\/a>/g;
    const snippetRe = /<a class="result__snippet"[^>]*>([\s\S]*?)<\/a>/g;
    const links: Array<{ url: string; title: string }> = [];
    let m: RegExpExecArray | null;
    while ((m = linkRe.exec(html)) && links.length < 5) {
      // DDG wraps the URL in a redirect; extract the actual URL
      const raw = m[1];
      const u = raw.match(/uddg=([^&]+)/);
      const actualUrl = u ? decodeURIComponent(u[1]) : raw;
      const title = m[2].replace(/<[^>]+>/g, '').trim();
      links.push({ url: actualUrl, title });
    }
    const snippets: string[] = [];
    while ((m = snippetRe.exec(html)) && snippets.length < 5) {
      snippets.push(m[1].replace(/<[^>]+>/g, '').trim());
    }
    for (let i = 0; i < links.length; i++) {
      results.push({ ...links[i], snippet: snippets[i] || '' });
    }
    if (results.length === 0) {
      return JSON.stringify({ error: 'no results found', query: args.query });
    }
    return JSON.stringify({ query: args.query, results });
  } catch (e: unknown) {
    const err = e as Error;
    return JSON.stringify({ error: err.message || String(e) });
  } finally {
    clearTimeout(timeout);
  }
}

File: discord-bot-fleet/bot-python/tools/web_search.py
"""web_search — DuckDuckGo HTML scrape, no API key required."""

import json
import re
from urllib.parse import quote
import aiohttp

WEB_SEARCH_TOOL = {
    'type': 'function',
    'function': {
        'name': 'web_search',
        'description': 'Search the web for current information. Returns up to 5 results with title, URL, and a short snippet. Use this when you need up-to-date info beyond your training data.',
        'parameters': {
            'type': 'object',
            'properties': {
                'query': {'type': 'string', 'description': 'Search query'},
            },
            'required': ['query'],
        },
    },
}


async def web_search_handler(args, ctx):
    query = args.get('query', '').strip()
    if not query:
        return json.dumps({'error': 'query required'})
    url = f'https://html.duckduckgo.com/html/?q={quote(query)}'
    try:
        async with aiohttp.ClientSession(
            timeout=aiohttp.ClientTimeout(total=10),
            headers={'User-Agent': 'Mozilla/5.0 (compatible; DiscordBot/1.0)'},
        ) as session:
            async with session.get(url) as resp:
                if resp.status != 200:
                    return json.dumps({'error': f'HTTP {resp.status}'})
                html = await resp.text()
        # Parse result blocks
        link_re = re.compile(r'<a class="result__a"[^>]*href="([^"]+)"[^>]*>([\s\S]*?)</a>', re.IGNORECASE)
        snippet_re = re.compile(r'<a class="result__snippet"[^>]*>([\s\S]*?)</a>', re.IGNORECASE)
        links = []
        for m in link_re.finditer(html):
            if len(links) >= 5:
                break
            raw = m.group(1)
            u = re.search(r'uddg=([^&]+)', raw)
            actual_url = __import__('urllib.parse').unquote(u.group(1)) if u else raw
            title = re.sub(r'<[^>]+>', '', m.group(2)).strip()
            links.append({'url': actual_url, 'title': title})
        snippets = []
        for m in snippet_re.finditer(html):
            if len(snippets) >= 5:
                break
            snippets.append(re.sub(r'<[^>]+>', '', m.group(1)).strip())
        results = []
        for i, link in enumerate(links):
            results.append({**link, 'snippet': snippets[i] if i < len(snippets) else ''})
        if not results:
            return json.dumps({'error': 'no results found', 'query': query})
        return json.dumps({'query': query, 'results': results})
    except Exception as e:
        return json.dumps({'error': str(e)})
